proposals/0002-testing-log-handler.md

```diff
@@ -0,0 +1,112 @@
+# Test Log Handler
+
+Authors: [Brian Maher](https://github.com/brimworks), [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+
+## Introduction
+
+Software operators rely on logging to provide insight into what is being done,
+therefore, testing log output is an important facet.
+
+## Motivation
+
+A variety of implementations have cropped up organically, but having a single
+test log handler to build off of will reduce the re-write churn.
+
+### Existing Implementations
+
+* [swift-log-testing](https://github.com/neallester/swift-log-testing)
+* [swift-distributed-actors/LogCapture](https://github.com/apple/swift-distributed-actors/blob/main/Sources/DistributedActorsTestKit/LogCapture.swift)
+* [swift-cluster-membership/LogCapture](https://github.com/apple/swift-cluster-membership/blob/main/Tests/SWIMTestKit/LogCapture.swift)
+* [RediStack/RedisLoggingTests](https://github.com/swift-server/RediStack/blob/master/Tests/RediStackIntegrationTests/RedisLoggingTests.swift)
+* [swift-log/TestLogger](https://github.com/apple/swift-log/blob/main/Tests/LoggingTests/TestLogger.swift#L52)
+
+
+## Features
+
+A testing log handler needs the following features:
+* The ability to capture log entries at the start of a unit test.
+* The ability to assert the captured log entries contain a matching message.
+* Work correctly in the face of parallel test execution.
+* Print log messages to the console if the log entry does not match a log entry
+  that is being asserted.
+* Bootstrap the test log handler as early as possible so we don't miss any
+  `Logger()` creations which may happen during setup.
+
+## Proposed Solution
+
+A global function to bootstrap the test log handler:
+
+```swift
+TestingLogHandler.bootstrap()
+```
+
+A way to create a new "container" which will capture all log messages that match
+a filter. Note that multiple concurrent "container"s need to collect duplicate
+log messages, since tests may be ran in parallel.
+
+To create a new "container" we can have a function that takes a filter parameter
+and returns the new log container:
+
+```swift
+let container = TestingLogHandler.container { logMessage in
+    // Ideally, you would be as specific as possible in your filtering!
+    return logMessage.level == Logger.Level.debug &&
+        logMessage.label == "Test" &&
+        logMessage.message.description == "Matched"
+}
+// Emit a log that matches the filter:
+Logger(label: "Test").debug("Matched")
+
+// Emit a log that does NOT match the filter:
+Logger(label: "Test").debug("Printed to console")
+
+// Assert that the log message was found:
+XCTAssertFalse(container.messages.isEmpty)
+```
+
+It is recommended to create a separate container for each log assertion so that
+you can simply check that the messages in a container are non-empty. If you use
+this pattern you also protect yourself from accidental log message matches that
+occur due to a test ran in parallel. An accidental match could cause a failing
```

Comment:
How so? The testing log handler is global, we may get multiple containers but it concurrently logging from various tests/threads, won't they end up in all currently active containers?

Reply:
>  The testing log handler is global, we may get multiple containers but it concurrently logging from various tests/threads, won't they end up in all currently active containers?

Yes that is correct understanding. Maybe I need to elaborate a bit more here... but the scenario that I'm thinking of is where test A causes log message X to be printed and test B also causes log message X to be printed. If test A and B are ran such that the log containers for test A and B are both created before log message X is printed, and the log message X is printed before test A and B assert the container is non-empty, there could be multiple messages in the container. Therefore asserting that there is exactly one message could lead to spurious errors.

Reply:
I think I'm more in favor of giving more powerful assertion capabilities that solve the filtering, rather than having a robust filtering API, because the former is less work for the library to build, test, and maintain, than the filtering capabilities of the log handler.

It's also more familiar territory for those looking for test utilities, as it's a bit more discoverable due to the fact you're reaching for these tools as you're drafting your assertions rather than when you're in the mindset of bootstrapping the test (at least to me)

Reply:
Thanks for the feeback @Mordil!

> because the former (more powerful assertion capabilities) is less work for the library to build, test, and maintain, than the filtering capabilities of the log handler.

Really? I would think it would be the opposite. More specifically, any assertion capability needs to have the exact same logic as the filtering capabilities... since an assertion is really just a fancy "filter" that not only returns true/false built in the event of failure it would print a nice debug message to assist.

> It's also more familiar territory for those looking for test utilities, as it's a bit more discoverable due to the fact you're reaching for these tools as you're drafting your assertions rather than when you're in the mindset of bootstrapping the test (at least to me)

Maybe the real problem is with the names I have chosen. Instead of having a "container" that has a pre-filter as I have done so far. What if we reduce the capabilities and instead have the exact same API but call this a "future log assertion"... and then later in the test you simply assert that this "future log assertion" got called (and remove the capability of introspecting any log messages that got captured). This would be similar to many "mock" frameworks which allow one to validate that certain methods on the "mock" got called. Do you think that would be more intuitive and discoverable?

Reply:
Obviously, I'll use whatever is settled on, and I think there's merit to both approaches.

My main feedback is just try to find the solution that requires the least amount of effort, has the least room for bugs, and perhaps has the most extensibility _outside_ of the module so this library doesn't need to see frequent bug/feature requests on this aspect.
---
I guess it's one of those "I'll know it if I see it" scenarios for me.

When I see the up-front filtering, I'm jumping (and this is where I may be wrong 😅) to the idea of an implementation that needs to track all of that, with a "nice" API for describing the filtering.

Vs. the assertions route is much more language primitives where we just store them, and then APIs for exposing those data structures and then relying on built-in methods like filter, map, etc.

In the former, I'd think you'd need more internal data structures, more internal logic. While in the latter, I'm thinking it's more of just a plain data structure of arrays.

Reply:
@Mordil  are you willing to put a little effort in and try out the implementation I have proposed? If you want to, you can have the pre-filter match all logs. When that container is active, no longs will be printed to stderr.

```swift
class TestMyLogs: XCTestCase {
    override func setUp() {
        TestingLogHandler.bootstrap() 
    }
    func testExample() {
        // Logger may be created before the container:
        let logger = Logger(label: "Example")
        let container = TestingLogHandler.container { true }
        logger.debug("Message")
        XCTAssertTrue(container.messages.find { $0.match(label: "Example", level: .debug) }, "Expected log message with label: Example, level: debug, but got \(container.messages.map { "\($0)" }.joined("\n"))")
    }
}
```

...and you could aways add extensions to the `TestingLogHandler.Container` if you want to make it easier to write assertions like the one above.

So, I think what I currently propose is the minimal code with maximal flexibility.

Personally, I wouldn't use it in this way, but everyone has their own preferences :).